include/mqtt/client.hpp

```diff
@@ -506,8 +506,9 @@ class client : public endpoint<Socket, std::mutex, std::lock_guard, PacketIdByte
         auto it = eps.begin();
         auto end = eps.end();
 #endif // BOOST_VERSION < 106600
-        base::socket() = std::move(socket);
-        connect_impl(*base::socket(), it, end, std::move(props), std::move(func));
+        socket_ = std::move(socket);
+        base::socket_optional().emplace(socket_);
```

Comment:
It is the most difficult and controversy point of the PR I think.
Let me explain.

* The `optional` is only for lazy initialization. Once the socket is created then it would never be updated.
* The design is the same as the current master.
https://github.com/redboltz/mqtt_cpp/blob/757cba16da08018cce603f840d61f7824681debf/include/mqtt/endpoint.hpp#L7302-L7308
https://github.com/redboltz/mqtt_cpp/blob/757cba16da08018cce603f840d61f7824681debf/include/mqtt/client.hpp#L462-L475
https://github.com/redboltz/mqtt_cpp/blob/757cba16da08018cce603f840d61f7824681debf/include/mqtt/client.hpp#L778-L782
  * On TLS connection, user needs to setup `ssl::context` before connect. The socket requires `ssl::context` as the constructor parameter. So the socket should be created in the `connect()`, not in the `client` constructor. However, `endpoint` has the member of `shared_ptr<Scoket>`, so it is initialized as nullptr at construction time. Then lazy initialized in `connect()`. This is the original (master) design.

  * In order to do the same things, I introduced `optional` in this PR. Because `shared_any` cannot be initialized by nullptr. Because it requires a reference of the pointee that is hold by the shared_ptr. 

That is the reason I introduced `optional`.

Reply:
Thank you for the explanation.

I'd rather this change be done separately from the mqtt::buffer changes

But I understand that that is a lot of work to separate the two changes.

Thoughts?

Reply:
> I'd rather this change be done separately from the mqtt::buffer changes
But I understand that that is a lot of work to separate the two changes.

Yes it is difficult. I can't separate them. Because type erasure is required for zero copy buffer without increasing compile times. So I'd like to go this way.

BTW, API `socket_optional()` is protected. So ordinary user doesn't care about `optional`.

Reply:
Can they be separated into two separate commits, on the same PR?

Reply:
If we keep successful compile result, we can't do that. Because zero copy and not slow compile times required both.

Once I commit slower compile times version, and then modify it, it could be possible but it is hard to me. It is unnecessary work and might take a long time.
So I'd like to commit them in the same commit.

Reply:
Alright. 

I understand.

I'm still not sure what causes the increased compile times, which is why I wanted to see your code by itself, but I understand that it's extra work.

Reply:
I pushed the slow compile times branch that is not type erased.

It is not continuous to the current impl_320_2 branch but it might help to analyze the reason of compile times increasing.

https://github.com/redboltz/mqtt_cpp/tree/impl_320_no_te_slow_compile